File: internal/api/reconnect.go
package api

import (
	"bytes"
	"encoding/json"
	"fmt"
	"sync"
	"time"
)

// Incremental SSE reconnect. A client returning to a still-open tab already
// holds the hoard list; instead of re-streaming ~100k rows it sends
// ?since=<server_ts it last saw> and we reply with just the delta:
//   - added   torrents (full current row, looked up by info_hash)
//   - removed torrents (info_hash only)
// observed since that cursor. If we can't guarantee completeness for that
// cursor (too old / evicted / server restarted) we fall back to a full
// hydration.
//
// Both adds and removes are timestamped by SERVER observation time (when the
// event crossed the hub), never the torrent's self-reported added_time — that
// field is 0 for a freshly-added torrent, which is exactly the case a delta
// must catch.

type rcEvent struct {
	hash  string
	at    float64 // unix seconds, server-observed
	added bool    // true = torrent_added, false = torrent_removed
}

type reconnectState struct {
	mu    sync.Mutex
	ring  []rcEvent
	cap   int
	floor float64 // earliest unix time from which the delta is still complete
}

func newReconnectState() *reconnectState {
	return &reconnectState{cap: 16384, floor: float64(time.Now().Unix())}
}

func (r *reconnectState) record(hash string, added bool) {
	if hash == "" {
		return
	}
	now := float64(time.Now().Unix())
	r.mu.Lock()
	r.ring = append(r.ring, rcEvent{hash: hash, at: now, added: added})
	if len(r.ring) > r.cap {
		drop := len(r.ring) - r.cap
		r.floor = r.ring[drop-1].at // everything up to here is no longer guaranteed
		r.ring = append([]rcEvent(nil), r.ring[drop:]...)
	}
	r.mu.Unlock()
}

// changesSince returns adds/removes observed after `since`, and whether the
// delta is trustworthy (since within the guaranteed window).
func (r *reconnectState) changesSince(since float64) (added, removed []string, ok bool) {
	r.mu.Lock()
	defer r.mu.Unlock()
	if since < r.floor {
		return nil, nil, false
	}
	for _, e := range r.ring {
		if e.at <= since {
			continue
		}
		if e.added {
			added = append(added, e.hash)
		} else {
			removed = append(removed, e.hash)
		}
	}
	return added, removed, true
}

// startReconnectWatcher records torrent_added / torrent_removed off the hoard
// hub. A cheap byte prefilter skips the 1Hz stats frames without a JSON parse.
func (s *Server) startReconnectWatcher() {
	if s.hoardEngine == nil {
		return
	}
	hub := s.hoardEngine.EventHub()
	if hub == nil {
		return
	}
	mAdd := []byte(`"torrent_added"`)
	mRem := []byte(`"torrent_removed"`)
	go func() {
		_, ch := hub.Subscribe()
		for frame := range ch {
			isAdd := bytes.Contains(frame, mAdd)
			isRem := bytes.Contains(frame, mRem)
			if !isAdd && !isRem {
				continue
			}
			var env struct {
				Event string `json:"event"`
				Data  struct {
					InfoHash string `json:"info_hash"`
				} `json:"data"`
			}
			if json.Unmarshal(frame, &env) != nil {
				continue
			}
			switch env.Event {
			case "torrent_added":
				s.reconnect.record(env.Data.InfoHash, true)
			case "torrent_removed":
				s.reconnect.record(env.Data.InfoHash, false)
			}
		}
	}()
}

// streamDelta emits the incremental update for a reconnecting client. Returns
// false when a full hydration is required instead.
func (s *Server) streamDelta(w interface{ Write([]byte) (int, error) }, flusher interface{ Flush() }, since float64) bool {
	if s.reconnect == nil {
		return false
	}
	// The delta only tracks this node's own engines. With agents registered, a
	// torrent that arrived on one since the last connect would be missing from
	// the delta and stay invisible until a full reload, so decline and let the
	// caller hydrate in full -- which does aggregate agents.
	if len(s.agentsSnapshot()) > 0 {
		return false
	}
	added, removed, ok := s.reconnect.changesSince(since)
	if !ok {
		return false
	}
	// Materialize full rows for the added hashes still present in the list.
	var addRows []map[string]interface{}
	if len(added) > 0 && s.hoardEngine != nil {
		want := make(map[string]bool, len(added))
		for _, h := range added {
			want[h] = true
		}
		for _, t := range s.hoardEngine.GetTorrentList() {
			if h, _ := t["info_hash"].(string); want[h] {
				addRows = append(addRows, t)
			}
		}
	}
	rcWriteFrame(w, map[string]interface{}{"event": "sync", "data": map[string]interface{}{"mode": "delta"}})
	if len(addRows) > 0 {
		rcWriteFrame(w, map[string]interface{}{
			"event": "torrent_batch",
			"data":  map[string]interface{}{"mode": "hoard", "torrents": addRows, "done": true},
		})
	}
	for _, h := range removed {
		rcWriteFrame(w, map[string]interface{}{"event": "torrent_removed", "data": map[string]interface{}{"info_hash": h}})
	}
	flusher.Flush()
	return true
}

func rcWriteFrame(w interface{ Write([]byte) (int, error) }, obj map[string]interface{}) {
	if frame, err := json.Marshal(obj); err == nil {
		fmt.Fprintf(w, "data: %s\n\n", frame)
	}
}

File: typhon-engine/src/hydra/reconnect.rs
//! Letting a returning client keep its place.
//!
//! A tab that comes back already holds the library. Re-streaming it costs
//! 187 MB and 300k rows for a client that was away four seconds -- the "it
//! takes about seven seconds to come back to life" report. Instead the client
//! sends the cursor it left on and gets only what changed.
//!
//! Events are timestamped by SERVER observation time, when the change crossed
//! this process, and never by the torrent's own `added_time`: that field is 0
//! for a freshly added torrent, which is exactly the case a delta has to catch.

use std::collections::VecDeque;
use std::sync::Mutex;

/// One observed change.
#[derive(Debug, Clone)]
struct Event {
    hash: String,
    at: f64,
    /// true = the row must be (re)sent, false = the torrent is gone.
    ///
    /// Not "was added": a row whose category or trackers changed is replayed
    /// as an add, because the delta carries full rows for added hashes and
    /// that is exactly the refresh a returning client needs.
    added: bool,
}

/// The window of changes this process can still speak for.
pub struct Ring {
    inner: Mutex<Inner>,
}

struct Inner {
    events: VecDeque<Event>,
    cap: usize,
    /// Earliest time from which the delta is still complete. Older than this
    /// and we no longer know what was missed.
    floor: f64,
}

impl Default for Ring {
    fn default() -> Self {
        // The floor starts at now: this process has no history from before it
        // started, so a cursor older than its boot is refused and the client
        // reloads. That is the honest answer, not an empty delta.
        Self::with_floor(now())
    }
}

impl Ring {
    fn with_floor(floor: f64) -> Self {
        Self { inner: Mutex::new(Inner { events: VecDeque::new(), cap: 16384, floor }) }
    }
}

fn now() -> f64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_secs() as f64)
        .unwrap_or(0.0)
}

impl Ring {
    pub fn record(&self, hash: &str, added: bool) {
        self.record_at(hash, added, now());
    }

    /// The same, with the observation time given rather than read.
    ///
    /// The clock is an input here, not a dependency: a test that has to sleep
    /// to observe an ordering is a test that fails on a busy machine.
    fn record_at(&self, hash: &str, added: bool, at: f64) {
        if hash.is_empty() {
            return;
        }
        let mut inner = self.inner.lock().unwrap();
        inner.events.push_back(Event { hash: hash.to_string(), at, added });
        while inner.events.len() > inner.cap {
            // Everything up to the dropped event is no longer guaranteed: a
            // client whose cursor predates it must be told to reload in full.
            if let Some(dropped) = inner.events.pop_front() {
                inner.floor = dropped.at;
            }
        }
    }

    /// A row changed without being added or removed: a category set, tags
    /// edited, a tracker added.
    pub fn record_changed(&self, hashes: &[String]) {
        for h in hashes {
            self.record(h, true);
        }
    }

    /// What changed after `since`, and whether the answer can be trusted.
    ///
    /// `None` means the cursor is older than the window: the honest answer is
    /// "reload everything", not a delta that silently omits what was dropped.
    pub fn changes_since(&self, since: f64) -> Option<(Vec<String>, Vec<String>)> {
        let inner = self.inner.lock().unwrap();
        if since < inner.floor {
            return None;
        }
        let mut added = Vec::new();
        let mut removed = Vec::new();
        for e in inner.events.iter().filter(|e| e.at > since) {
            if e.added {
                added.push(e.hash.clone());
            } else {
                removed.push(e.hash.clone());
            }
        }
        Some((added, removed))
    }
}

/// Whether a delta may be served at all.
///
/// ⚠ The question is whether a DIALLED agent exists -- a node this process
/// reaches over the network, whose torrents its own hubs cannot speak for. A
/// torrent that arrived on one while the client was away would be missing from
/// the delta and stay invisible until a full reload.
///
/// This node's own engines are agents by name (`local-race`, `local-hoard`,
/// `local-race-2`, since one engine became one agent) and are NOT that case:
/// their rows are right here. Counting them as remote is what silently
/// disabled the delta on every multi-engine node -- the count was never zero
/// again, so every tab return re-streamed the whole library.
pub fn delta_allowed(agents: &[(String, bool)]) -> bool {
    agents.iter().all(|(_, is_local)| *is_local)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_cursor_inside_the_window_gets_only_what_changed() {
        let r = Ring::with_floor(100.0);
        r.record_at("aaa", true, 110.0);
        r.record_at("bbb", false, 120.0);
        let (added, removed) = r.changes_since(105.0).expect("the window covers this cursor");
        assert_eq!(added, ["aaa"]);
        assert_eq!(removed, ["bbb"]);
    }

    /// A change and a reconnect inside the same second: the timestamps are
    /// whole seconds, as 3.x's are, so `at > since` drops it. Faithful, and
    /// worth knowing -- the client sees it on its next event, not never.
    #[test]
    fn the_window_is_exclusive_on_its_lower_bound() {
        let r = Ring::with_floor(100.0);
        r.record_at("aaa", true, 110.0);
        assert!(r.changes_since(110.0).unwrap().0.is_empty());
        assert_eq!(r.changes_since(109.0).unwrap().0, ["aaa"]);
    }

    #[test]
    fn a_cursor_older_than_the_window_is_refused_rather_than_answered_partially() {
        let r = Ring::with_floor(1_000.0);
        assert!(
            r.changes_since(999.0).is_none(),
            "a delta that omits what was dropped is worse than asking for a reload"
        );
        assert!(r.changes_since(1_000.0).is_some());
    }

    #[test]
    fn an_edited_row_is_replayed_as_an_add() {
        // The delta carries full rows for added hashes, so replaying an edit
        // that way is the refresh the client needs -- it has no other channel
        // for "this row changed".
        let r = Ring::with_floor(100.0);
        r.record_at("aaa", true, 110.0);
        let (added, removed) = r.changes_since(105.0).unwrap();
        assert_eq!(added, ["aaa"]);
        assert!(removed.is_empty());
    }

    /// ⭐ The bug 3.181.0 fixed. Since one engine became one agent, a
    /// multi-engine node always has agents registered, so a guard counting
    /// them never fired -- and every tab return re-streamed 187 MB.
    #[test]
    fn this_nodes_own_engines_do_not_disable_the_delta() {
        let local_only = vec![
            ("local-race".to_string(), true),
            ("local-hoard".to_string(), true),
            ("local-race-2".to_string(), true),
        ];
        assert!(delta_allowed(&local_only), "our own engines are not a reason to give up");

        let with_remote = vec![("local-race".to_string(), true), ("far-node".to_string(), false)];
        assert!(
            !delta_allowed(&with_remote),
            "a dialled agent holds rows this process cannot speak for"
        );
        assert!(delta_allowed(&[]), "no agent at all is not a reason either");
    }
}
